atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunner.java

```diff
@@ -15,19 +15,361 @@
+                                                         Sweeper sweeper) {
+        if (sweeper.shouldSweepLastCommitted()) {
+            return keyValueService.getRange(tableRef, range, sweepTs);
+        } else {
+            return ClosableIterators.emptyImmutableClosableIterator();
+        }
+    }
+
+    public static Equivalence<RowResult<Set<Long>>> sameRowEquivalence() {
+        return new Equivalence<RowResult<Set<Long>>>() {
+            @Override
+            protected boolean doEquivalent(RowResult<Set<Long>> fst, RowResult<Set<Long>> snd) {
+                return Arrays.equals(fst.getRowName(), snd.getRowName());
+            }
+
+            @Override
+            protected int doHash(RowResult<Set<Long>> setRowResult) {
+                return 0;
+            }
+        };
+    }
+
+    private int sweepForCells(
+            CellsAndTimestamps currentBatch,
+            TableReference tableRef,
+            Sweeper sweeper,
+            long sweepTs,
+            PeekingIterator<RowResult<Value>> peekingValues,
+            RunType runType) {
+        CellsAndTimestamps currentBatchWithoutIgnoredTimestamps =
+                currentBatch.withoutIgnoredTimestamps(sweeper.getTimestampsToIgnore());
+
+        CellsToSweep cellsToSweep = getStartTimestampsPerRowToSweep(
+                currentBatchWithoutIgnoredTimestamps, peekingValues, sweepTs, sweeper);
+
+        Multimap<Cell, Long> startTimestampsToSweepPerCell = cellsToSweep.timestampsAsMultimap();
+
+        if (runType == RunType.FULL) {
+            cellsSweeper.sweepCells(tableRef, startTimestampsToSweepPerCell, cellsToSweep.allSentinels());
+        }
+
+        return startTimestampsToSweepPerCell.size();
+    }
+
+    private static Iterator<CellAndTimestamps> getTimestampsFromRowResultsIterator(
+            Iterable<RowResult<Set<Long>>> cellsToSweep) {
+        return StreamSupport.stream(cellsToSweep.spliterator(), false)
+                .flatMap(SweepTaskRunner::rowToCellAndTimestampStream)
+                .iterator();
+    }
+
+    private static Stream<CellAndTimestamps> rowToCellAndTimestampStream(RowResult<Set<Long>> rowResult) {
+        Set<Map.Entry<Cell, Set<Long>>> cellsInRow = ImmutableSet.copyOf(rowResult.getCells());
+        return cellsInRow.stream()
+                .map(SweepTaskRunner::convertToCellAndTimestamps);
+    }
+
+    private static CellAndTimestamps convertToCellAndTimestamps(
+            Map.Entry<Cell, Set<Long>> entry) {
+        return CellAndTimestamps.of(entry.getKey(), entry.getValue());
+    }
+
+    @VisibleForTesting
+    CellsToSweep getStartTimestampsPerRowToSweep(
+            CellsAndTimestamps startTimestampsPerCell,
+            PeekingIterator<RowResult<Value>> values,
+            long sweepTimestamp,
+            Sweeper sweeper) {
+
+        LoadingCache<Long, Long> startTsToCommitTs = CacheBuilder.newBuilder()
+                .build(new StartTsToCommitTsCacheLoader(transactionService));
```

Comment:
No. All timestamps we are going to load are already loaded in memory.

But again, not my code, this is just a rename.